lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -950,80 +911,103 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
 
     auto TH = llvm::TrackingMDNodeRef(FwdDecl.get());
     DITypeCache[DbgTy.getType()] = TH;
-    auto Members =
-        getStructMembers(Decl, BaseTy, Scope, File, Flags, SizeInBits);
+    // Collect the members.
+    SmallVector<llvm::Metadata *, 16> Elements;
+    unsigned OffsetInBits = 0;
+    for (VarDecl *VD : Decl->getStoredProperties()) {
+      auto memberTy =
+          BaseTy->getTypeOfMember(IGM.getSwiftModule(), VD, nullptr);
+
+      if (auto DbgTy = CompletedDebugTypeInfo::getFromTypeInfo(
+              VD->getInterfaceType(),
+              IGM.getTypeInfoForUnlowered(
+                  IGM.getSILTypes().getAbstractionPattern(VD), memberTy)))
+        Elements.push_back(createMemberType(*DbgTy, VD->getName().str(),
+                                            OffsetInBits, Scope, File, Flags));
+      else
+        // Without complete type info we can only create a forward decl.
```

Comment:
Do you know when this happens? This is unfortunate, because one "bad" member could silently virally infect other types, and this function essentially suppresses that knowledge.

Reply:
Typically this happens in function signatures, or really any cases where a type is nested inside another type and we don't have any insight into its storage size because no code has been generated for it.

At some point I think it might be good to revisit this all. We currently conflate the size of a type and the static size of the  storage allocated for a type. The latter is a fixed-sized container (e.g., a pointer for a class, ...) but we don't have any Swift or SIL type system nodes to represent containers, because they are just an artifact of IRGen. In DWARF we currently create the fixed size containers as anonymous structs with a size (provided by IRGen) and a single unsized element of the real type.

You are absolutely right that we should not cache a forward-declared type so that if we later encounter a full definition for it, we can replace it.

Reply:
> You are absolutely right that we should not cache a forward-declared type so that if we later encounter a full definition for it, we can replace it.

Which we already do:

Reply:
from getOrCreateType()
```
    // If this is a forward decl, create one for this mangled name and don't
    // cache it.
```